File: model/src/main/scala/ru/mipt/acsl/decode/model/types/Const.scala
package ru.mipt.acsl.decode.model.types

import ru.mipt.acsl.decode.model.naming.{ElementName, Namespace}

/**
  * @author Artem Shein
  */
trait Const extends DecodeType {

  def value: String

  override def toString: String =
    s"${this.getClass}{alias = $alias, namespace = $namespace, value = $value}"

}

object Const {

  private class ConstImpl(val alias: Option[Alias.NsConst], var namespace: Namespace, val value: String,
                          val typeParameters: Seq[ElementName])
    extends Const {

    def systemName: String = value

  }

  def apply(alias: Option[Alias.NsConst], ns: Namespace, value: String): Const =
    new ConstImpl(alias, ns, value, Seq.empty)

}

File: model/src/main/java/ru/mipt/acsl/decode/model/types/Const.java
package ru.mipt.acsl.decode.model.types;

import org.jetbrains.annotations.Nullable;
import ru.mipt.acsl.decode.model.naming.ElementName;
import ru.mipt.acsl.decode.model.naming.Namespace;

import java.util.List;

/**
 * Created by metadeus on 07.06.16.
 */
public interface Const extends DecodeType {

    static Const newInstance(@Nullable Alias.NsConst alias, Namespace namespace, String value,
                                                   List<ElementName> typeParameters) {
        return new ConstImpl(alias, namespace, value, typeParameters);
    }

    String value();

}
